unit/analyses/ai/ai.cpp

```diff
@@ -0,0 +1,310 @@
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+/// \file
+/// Unit tests for selectively retaining particular domains in ait
+
+#include <testing-utils/catch.hpp>
+
+#include <analyses/ai.h>
+
+#include <ansi-c/ansi_c_language.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+#include <langapi/mode.h>
+
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/c_types.h>
+
+/// A very simple analysis that counts executed instructions along a particular
+/// path, taking the max at merge points and saturating at 100 instructions.
+/// It should indicate that instructions not within a loop have a certain path
+/// length, and those reachable from a loop have a path length of 100
+/// (i.e. potentially infinity)
+class instruction_counter_domaint : public ai_domain_baset
+{
+public:
+
+  std::unique_ptr<int> path_length;
+
+  instruction_counter_domaint()
+  {
+  }
+
+  instruction_counter_domaint(const instruction_counter_domaint &other)
+  {
+    if(other.path_length)
+      path_length.reset(new int(*other.path_length));
+  }
+
+  void transform(locationt, locationt, ai_baset &, const namespacet &) override
+  {
+    if(*path_length < 100)
+      ++*path_length;
+  }
+
+  void make_bottom() override
+  {
+    path_length.reset();
+  }
+  void make_top() override
+  {
+    UNREACHABLE;
+  }
+  void make_entry() override
+  {
+    path_length.reset(new int);
+    *path_length = 0;
+  }
+  bool is_bottom() const override
+  {
+    return path_length == nullptr;
+  }
+  bool is_top() const override
+  {
+    UNREACHABLE;
+    return true;
+  }
+
+  bool merge(const instruction_counter_domaint &b, locationt from, locationt to)
+  {
+    if(b.is_bottom())
+      return false;
+
+    if(!path_length)
+    {
+      path_length.reset(new int);
+      *path_length = 0;
+    }
+
+    int old_count = *path_length;
+    // Max path length to get here:
+    *path_length =
+      std::max(*path_length, *b.path_length);
+    return *path_length != old_count;
+  }
+};
+
+class instruction_counter_analysist : public ait<instruction_counter_domaint>
+{
+public:
+
+  static bool is_y_assignment(const goto_programt::instructiont &i)
+  {
+    if(!i.is_assign())
+      return false;
+    const code_assignt &assign = to_code_assign(i.code);
+    return
+      assign.lhs().id() == ID_symbol &&
+      id2string(to_symbol_expr(assign.lhs()).get_identifier()).find('y') !=
+      std::string::npos;
+  }
+
+  static bool is_y_assignment_location(locationt l)
+  {
+    // At assignments to variables with a 'y' in their name, keep the domain.
+    // Otherwise let ait decide whether to keep it.
+    return is_y_assignment(*l);
+  }
+};
+
+static code_function_callt make_void_call(const symbol_exprt &function)
+{
+  code_function_callt ret;
+  ret.function() = function;
+  return ret;
+}
+
+SCENARIO(
+  "ait general testing", "[core][analyses][ai][general]")
+{
+  // Make a program like:
```

Comment:
I don't have a better suggestion but I worry this might be a bit fragile w.r.t. changes / clarification / refinement of what it means for a program to be well structured.